Add count support for listing filters

davelopez opened the pull request:
Extracted from https://github.com/galaxyproject/galaxy/pull/16003

I figured out that could be useful when trying to get the `total_matches` of a query using the current "dynamic filtering framework", not sure why it was not part of the API already, maybe because using functional filters could impact the performance in some extreme cases? We probably could fail the request as we do currently when listing history contents.

In draft until both TODOs are resolved.

## How to test the changes?
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Add count support for listing filters

Allows to get the total_matches of a query (without offset or limit) to be able to handle pagination in a deterministic way.

mvdbeek commented on test/unit/app/managers/test_HistoryManager.py:
```diff
@@ -938,6 +938,38 @@ def test_list(self):
+        self.log("orm filtered should work")
+        filters = [model.History.deleted == true()]
+        assert self.history_manager.count(filters=filters) == len(deleted)
+        self.log("fn filtered should work")
```
MInor nitpick, but if you find yourself having to do this it's probably better to break up the test into multiple tests.

davelopez replied:
I did it like this just because all the other tests were like this... but yeah I felt dirty while doing it :laughing: 
I can refactor all of them to break them down into smaller tests as part of this enhancement, although there are 106 of them just in this file... maybe better just turn the log statements into comments? :sweat_smile:

davelopez replied:
I turn them into assert messages to make them a bit more useful :+1:

mvdbeek commented on lib/galaxy/managers/base.py:
```diff
@@ -387,6 +387,27 @@ def list(self, filters=None, order_by=None, limit=None, offset=None, **kwargs):
+
+        # TODO: raise an error instead to avoid performance issues?
+
+        # fn filters will change the number of items
```
The list of known function filters is `annotation`, `genome_build`, `data_type` and `community_rating`. `community_rating` we should drop, and the rest seems like it could be converted into orm or column filters relatively easily ... or we should at least give it a try.

In the meantime I think just raising an invalid request exception seems fine.

davelopez replied:
Thank you! I will do that

davelopez committed:
Do not support functional filters in count

The existing functional filters should be either translated to ORM filters
or dropped. They won't be supported in count.

davelopez committed:
Turn test logs into assert messages

This test file is full of them, and it is probably better to split all of them into their own test cases but it will require a bigger refactoring...

davelopez committed:
Handle case insensitive filters on count
